packages/btcstaking/src/staking.rs

```diff
@@ -0,0 +1,830 @@
+    let calculated_fee = staking_output_value - total_output_value;
+    if calculated_fee < slashing_tx_min_fee {
+        return Err(Error::InsufficientSlashingFee(slashing_tx_min_fee));
+    }
+
+    Ok(())
+}
+
+/// Validates all relevant data of slashing and funding transactions.
+#[allow(clippy::too_many_arguments)]
+pub fn check_slashing_tx_match_funding_tx(
+    slashing_tx: &Transaction,
+    funding_transaction: &Transaction,
+    funding_output_idx: u32,
+    slashing_tx_min_fee: u64,
+    slashing_rate: f64,
+    slashing_pk_script: &[u8],
+    staker_pk: &VerifyingKey,
+    slashing_change_lock_time: u16,
+) -> Result<()> {
+    // Check if slashing tx min fee is valid
+    if slashing_tx_min_fee == 0 {
+        return Err(Error::InsufficientSlashingFee(0));
+    }
+
+    // Check if slashing rate is in the valid range (0,1)
+    if !is_rate_valid(slashing_rate) {
+        return Err(Error::InvalidSlashingRate {});
+    }
+
+    if funding_output_idx >= funding_transaction.output.len() as u32 {
+        return Err(Error::InvalidFundingOutputIndex(
+            funding_output_idx,
+            funding_transaction.output.len(),
+        ));
+    }
+
+    let staking_output = &funding_transaction.output[funding_output_idx as usize];
+
+    // Check if slashing transaction is valid
+    validate_slashing_tx(
+        slashing_tx,
+        slashing_pk_script,
+        slashing_rate,
+        slashing_tx_min_fee,
+        staking_output.value.to_sat(),
+        staker_pk,
+        slashing_change_lock_time,
+    )?;
+
+    // Check that slashing transaction input is pointing to staking transaction
+    let staking_tx_hash = funding_transaction.compute_txid(); // Hash of the funding transaction
+    if slashing_tx.input[0]
+        .previous_output
+        .txid
+        .ne(&staking_tx_hash)
+    {
+        return Err(Error::StakingOutputNotSpentBySlashingTx {});
+    }
+
+    // Check that index of the funding output matches index of the input in slashing transaction
+    if slashing_tx.input[0].previous_output.vout != funding_output_idx {
+        return Err(Error::StakingOutputNotSpentBySlashingTx {});
+    }
+
+    Ok(())
+}
+
+fn calc_sighash(
+    transaction: &Transaction,
+    funding_output: &TxOut,
+    path_script: &Script,
+) -> Result<[u8; 32]> {
+    // Check for incorrect input count
+    if transaction.input.len() != 1 {
+        return Err(Error::TxInputCountMismatch(1, transaction.input.len()));
+    }
+
+    // calculate tap leaf hash for the given path of the script
+    let tap_leaf_hash = path_script.tapscript_leaf_hash();
+
+    // calculate the sig hash of the tx with the given funding output
+    let mut sighash_cache = SighashCache::new(transaction);
+    let sighash = sighash_cache
+        .taproot_script_spend_signature_hash(
+            0,
+            &Prevouts::All(&[funding_output]),
+            tap_leaf_hash,
+            bitcoin::TapSighashType::Default,
+        )
+        .unwrap();
```

Comment:
The unwrap can be removed with a trivial change.

```suggestion
        )?;
```

```diff
diff --git a/packages/btcstaking/src/errors.rs b/packages/btcstaking/src/errors.rs
index ce64d29..7eab699 100644
--- a/packages/btcstaking/src/errors.rs
+++ b/packages/btcstaking/src/errors.rs
@@ -54,4 +54,6 @@ pub enum Error {
     TxHasSignatureScript {},
     #[error("Slashing or staking transaction values must be larger than 0")]
     InvalidSlashingAmount {},
+    #[error(transparent)]
+    SighashTaproot(#[from] bitcoin::sighash::TaprootError)
 }
 ```